File: src/SevenZipUtils.cpp
#include "SevenZipUtils.h"

namespace SevenZipUtils
{
    static std::wstring s_cachedInstallPath;
    static bool s_pathCached = false;

    static std::wstring ReadRegistryString(HKEY hRoot, LPCWSTR subKey, LPCWSTR valueName)
    {
        HKEY hKey = nullptr;
        if (RegOpenKeyExW(hRoot, subKey, 0, KEY_READ, &hKey) != ERROR_SUCCESS)
            return L"";

        WCHAR buffer[MAX_PATH] = {};
        DWORD size = sizeof(buffer);
        DWORD type = 0;
        LSTATUS status = RegQueryValueExW(hKey, valueName, nullptr, &type,
                                           reinterpret_cast<LPBYTE>(buffer), &size);
        RegCloseKey(hKey);

        if (status == ERROR_SUCCESS && (type == REG_SZ || type == REG_EXPAND_SZ))
            return std::wstring(buffer);
        return L"";
    }

    static bool FileExists(const std::wstring& path)
    {
        DWORD attrs = GetFileAttributesW(path.c_str());
        return (attrs != INVALID_FILE_ATTRIBUTES) && !(attrs & FILE_ATTRIBUTE_DIRECTORY);
    }

    std::wstring GetInstallPath()
    {
        if (s_pathCached)
            return s_cachedInstallPath;

        s_pathCached = true;

        // Try registry locations
        const LPCWSTR regPaths[] = {
            L"SOFTWARE\\7-Zip",
        };
        const LPCWSTR valueNames[] = { L"Path64", L"Path" };
        const HKEY roots[] = { HKEY_LOCAL_MACHINE, HKEY_CURRENT_USER };

        for (HKEY root : roots)
        {
            for (LPCWSTR regPath : regPaths)
            {
                for (LPCWSTR val : valueNames)
                {
                    std::wstring path = ReadRegistryString(root, regPath, val);
                    if (!path.empty())
                    {
                        // Ensure trailing backslash
                        if (path.back() != L'\\')
                            path += L'\\';

                        if (FileExists(path + L"7z.exe"))
                        {
                            s_cachedInstallPath = path;
                            return s_cachedInstallPath;
                        }
                    }
                }
            }
        }

        // Try well-known paths
        const LPCWSTR knownPaths[] = {
            L"C:\\Program Files\\7-Zip\\",
            L"C:\\Program Files (x86)\\7-Zip\\",
        };

        for (LPCWSTR p : knownPaths)
        {
            if (FileExists(std::wstring(p) + L"7z.exe"))
            {
                s_cachedInstallPath = p;
                return s_cachedInstallPath;
            }
        }

        return L"";
    }

    std::wstring Get7zExePath()
    {
        std::wstring dir = GetInstallPath();
        return dir.empty() ? L"" : dir + L"7z.exe";
    }

    std::wstring Get7zGUIPath()
    {
        std::wstring dir = GetInstallPath();
        return dir.empty() ? L"" : dir + L"7zG.exe";
    }

    std::wstring Get7zFMPath()
    {
        std::wstring dir = GetInstallPath();
        return dir.empty() ? L"" : dir + L"7zFM.exe";
    }

    bool IsArchiveExtension(const std::wstring& extension)
    {
        static const std::unordered_set<std::wstring> archiveExts = {
            L".7z", L".zip", L".rar", L".tar", L".gz", L".gzip", L".tgz",
            L".bz2", L".bzip2", L".tbz2", L".tbz", L".xz", L".txz",
            L".lzma", L".lz", L".zst", L".zstd",
            L".cab", L".iso", L".img",
            L".wim", L".swm", L".esd",
            L".arj", L".cpio", L".rpm", L".deb",
            L".lzh", L".lha",
            L".dmg", L".hfs",
            L".vhd", L".vhdx", L".vmdk", L".vdi",
            L".fat", L".ntfs", L".ext", L".ext2", L".ext3", L".ext4",
            L".squashfs", L".cramfs",
            L".z", L".taz",
            L".jar", L".war", L".ear",
            L".apk", L".aab",
            L".xpi", L".crx",
            L".nupkg",
            L".001",
            L".zipx",
        };

        std::wstring lower = extension;
        std::transform(lower.begin(), lower.end(), lower.begin(), ::towlower);
        return archiveExts.count(lower) > 0;
    }

    bool HasArchiveItem(IShellItemArray* psiItemArray)
    {
        if (!psiItemArray) return false;

        DWORD count = 0;
        psiItemArray->GetCount(&count);

        for (DWORD i = 0; i < count; i++)
        {
            ComPtr<IShellItem> item;
            if (SUCCEEDED(psiItemArray->GetItemAt(i, &item)))
            {
                PWSTR name = nullptr;
                if (SUCCEEDED(item->GetDisplayName(SIGDN_FILESYSPATH, &name)))
                {
                    std::wstring path(name);
                    CoTaskMemFree(name);

                    std::filesystem::path p(path);
                    if (IsArchiveExtension(p.extension().wstring()))
                        return true;
                }
            }
        }
        return false;
    }

    std::wstring GetIconString()
    {
        std::wstring dir = GetInstallPath();
        if (dir.empty()) return L"";
        return dir + L"7z.dll,0";
    }
}

File: src/7ZipMenu/SevenZipUtils.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using Microsoft.Win32;

namespace SevenZipMenu;

internal static class SevenZipUtils
{
    private static string? s_cachedInstallPath;
    private static bool s_pathCached;

    private static readonly HashSet<string> ArchiveExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".7z", ".zip", ".rar", ".tar", ".gz", ".gzip", ".tgz",
        ".bz2", ".bzip2", ".tbz2", ".tbz", ".xz", ".txz",
        ".lzma", ".lz", ".zst", ".zstd",
        ".cab", ".iso", ".img",
        ".wim", ".swm", ".esd",
        ".arj", ".cpio", ".rpm", ".deb",
        ".lzh", ".lha",
        ".dmg", ".hfs",
        ".vhd", ".vhdx", ".vmdk", ".vdi",
        ".fat", ".ntfs", ".ext", ".ext2", ".ext3", ".ext4",
        ".squashfs", ".cramfs",
        ".z", ".taz",
        ".jar", ".war", ".ear",
        ".apk", ".aab",
        ".xpi", ".crx",
        ".nupkg",
        ".001",
        ".zipx",
        ".xpi",
        ".docx", ".xlsx", ".pptx",
    };

    public static string GetInstallPath()
    {
        if (s_pathCached)
            return s_cachedInstallPath ?? string.Empty;

        s_pathCached = true;

        string[] valueNames = ["Path64", "Path"];
        RegistryKey[] roots = [Registry.LocalMachine, Registry.CurrentUser];

        foreach (var root in roots)
        {
            foreach (var valueName in valueNames)
            {
                string? path = ReadRegistryString(root, @"SOFTWARE\7-Zip", valueName);
                if (!string.IsNullOrEmpty(path))
                {
                    if (!path.EndsWith('\\'))
                        path += '\\';
                    if (File.Exists(path + "7z.exe"))
                    {
                        s_cachedInstallPath = path;
                        return s_cachedInstallPath;
                    }
                }
            }
        }

        string[] programFolders =
        [
            Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles),
            Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
        ];

        foreach (var folder in programFolders)
        {
            if (string.IsNullOrEmpty(folder)) continue;
            string candidate = Path.Combine(folder, "7-Zip") + "\\";
            if (File.Exists(candidate + "7z.exe"))
            {
                s_cachedInstallPath = candidate;
                return s_cachedInstallPath;
            }
        }

        return string.Empty;
    }

    public static string Get7zExePath()
    {
        string dir = GetInstallPath();
        return dir.Length == 0 ? string.Empty : dir + "7z.exe";
    }

    public static string Get7zGUIPath()
    {
        string dir = GetInstallPath();
        return dir.Length == 0 ? string.Empty : dir + "7zG.exe";
    }

    public static string Get7zFMPath()
    {
        string dir = GetInstallPath();
        return dir.Length == 0 ? string.Empty : dir + "7zFM.exe";
    }

    public static bool IsArchiveExtension(string extension) =>
        ArchiveExtensions.Contains(extension);

    public static bool HasArchiveItem(IShellItemArray? psia)
    {
        if (psia is null) return false;
        try
        {
            psia.GetCount(out uint count);
            for (uint i = 0; i < count; i++)
            {
                psia.GetItemAt(i, out IShellItem item);
                item.GetDisplayName(SIGDN.SIGDN_FILESYSPATH, out nint namePtr);
                string path = Marshal.PtrToStringUni(namePtr) ?? string.Empty;
                Marshal.FreeCoTaskMem(namePtr);

                string ext = Path.GetExtension(path);
                if (!string.IsNullOrEmpty(ext) && IsArchiveExtension(ext))
                    return true;
            }
        }
        catch { }
        return false;
    }

    public static string GetIconString()
    {
        string dir = GetInstallPath();
        return dir.Length == 0 ? string.Empty : dir + "7z.dll,0";
    }

    private static string? ReadRegistryString(RegistryKey root, string subKey, string valueName)
    {
        try
        {
            using var key = root.OpenSubKey(subKey);
            return key?.GetValue(valueName) as string;
        }
        catch { return null; }
    }
}
